tools/crmadmin.c

```diff
@@ -350,101 +320,124 @@ list_nodes()
-                break;
-            case 'B':
-                BASH_EXPORT = TRUE;
-                break;
-            case 'K':
-                command = cmd_shutdown;
-                crm_trace("Option %c => %s", flag, optarg);
-                if (dest_node != NULL) {
-                    free(dest_node);
-                }
-                dest_node = strdup(optarg);
-                break;
-            case 'q':
-                BE_SILENT = TRUE;
-                break;
-            case 'P':
-                command = cmd_pacemakerd_health;
-                need_pacemakerd_api = true;
-                need_controld_api = false;
-                break;
-            case 'S':
-                command = cmd_health;
-                crm_trace("Option %c => %s", flag, optarg);
-                if (dest_node != NULL) {
-                    free(dest_node);
-                }
-                dest_node = strdup(optarg);
-                break;
-            case 'E':
-                command = cmd_elect_dc;
-                break;
-            case 'N':
-                command = cmd_list_nodes;
-                need_controld_api = false;
-                break;
-            case 'H':
-                fprintf(stderr, "Cluster-wide health option not supported\n");
-                ++argerr;
-                break;
-            default:
-                printf("Argument code 0%o (%c) is not (?yet?) supported\n", flag, flag);
-                ++argerr;
-                break;
+    if (args->version) {
+        /* FIXME:  When crmadmin is converted to use formatted output, this can go. */
+        pcmk__cli_help('v', CRM_EX_USAGE);
+    }
+
+    if (options.timeout) {
+        message_timeout_ms = options.timeout;
+        if (message_timeout_ms < 1) {
+            message_timeout_ms = DEFAULT_MESSAGE_TIMEOUT_MS;
         }
     }
 
-    if (optind < argc) {
-        printf("non-option ARGV-elements: ");
-        while (optind < argc)
-            printf("%s ", argv[optind++]);
-        printf("\n");
+    if (options.dc_lookup) {
+        command = cmd_whois_dc;
+    }
+
+    if (options.kill) {
+        command = cmd_shutdown;
+        crm_trace("Option %c => %s", 'K', options.kill);
+        if (dest_node != NULL) {
+            free(dest_node);
+        }
+        dest_node = strdup(options.kill);
+    }
+
+    if (options.quiet) {
+        BE_SILENT = TRUE;
+    }
+
+    if (options.pacemakerd) {
+        command = cmd_pacemakerd_health;
+        need_pacemakerd_api = true;
+        need_controld_api = false;
+    }
+
+    if (options.status) {
+        command = cmd_health;
+        crm_trace("Option %c => %s", 'S', options.status);
+        if (dest_node != NULL) {
+            free(dest_node);
+        }
+        dest_node = strdup(options.status);
+    }
+
+    if (options.election) {
+        command = cmd_elect_dc;
+    }
+
+    if (options.nodes) {
+        command = cmd_list_nodes;
+        need_controld_api = false;
+    }
```

Comment:
The way you've done it is perfectly fine, but to help you get familiar with GOption's more advanced stuff, let's redo the command options.

You've seen how GOption can set basic argument types. When more complicated handling is needed, we can use G_OPTION_ARG_CALLBACK instead of a basic type. The value set in the option entry is then the name of a function that will be called when the option is parsed. See new_string_cb() in crm_diff and command_cb() in crm_node as examples.

In crmadmin, we can use a callback to set command and the need_* variables when the option is parsed, rather than afterward like here, and avoid some extra variables in the options struct. You can use a separate callback per command option, or one for all of them.

That's not necessarily better than your approach, but it'll get you comfortable with option callbacks.